### PR TITLE
Edited plotting file to cater for all the optimizers. Added a main fu…

### PR DESCRIPTION
**Pull request by umarsajjad1992**
## Purpose
Update the plotting file for optimization history to cater for all the optimizers in the current version

## Description
- The plotting file only catered for SNOPT optimizer and would give errors when using SLSQP
- An issue was raised and a suggestion of a PR to correct this behavior
- Add a parser argument with opt option to check for optimizer and then run the code accordingly
- Added an if else statement to check which optimizer is being used
- The code would change the layout and number of the plots accordingly
- Added a try and except block to check whether an optimization history file exists or not

## Expected time until merged
The PR is not urgent as, but a week should be enough for review and merging.

## Type of change
What types of change is it?
Select the appropriate type(s) that describe this PR

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
run the code:

`python plot_optHist.py --histFile output/opt.hst --outputDir output/ --opt "SNOPT"`

try changing the --opt argument to either "SLSQP" or "IPOPT"

## Checklist

- [x] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [ ] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Review by marcomangano — COMMENTED** [review]
Thanks for opening this PR! Looking good, I was initially confused by the choice of a large `try...except` block but I see you are gracefully exiting at the end, makes sense.
I will shortly push a commit to fix the black and flake8 errors. @hajdik it looks like an ESP test is failing, can you check what is going on?

File: tutorial/opt/aero/plot_optHist.py
Comment (ewu63) [inline tutorial/opt/aero/plot_optHist.py:22]: Why do you need this try block? Why can't you just check with `os.path.isfile`?

File: tutorial/opt/aero/plot_optHist.py
Comment (ewu63) [inline tutorial/opt/aero/plot_optHist.py:9]: These typically live outside of `main()`

File: tutorial/opt/aero/plot_optHist.py
Comment (ewu63) [inline tutorial/opt/aero/plot_optHist.py:18]: The user shouldn't need to specify the optimizer. It is stored in the metadata field of the history file, which you can get with `hist.getMetadata()`, which is a dict containing a field called `optimizer`.

File: tutorial/opt/aero/plot_optHist.py
Comment (ewu63) [inline tutorial/opt/aero/plot_optHist.py:29]: It seems most of the plotting is the same between the two blocks, only difference is SNOPT has feasibility/optimality in `axes[0]` and there are 5 subplots. You can probably consolidate this so there is no code duplication. E.g.

```
if SNOPT:
    # make 5 subplots
    # plot optimality/feasibility on axes[00
else:
    # make 4 subplots
# plot the rest on axes[-4:]
```

File: tutorial/opt/aero/plot_optHist.py
Comment (marcomangano) [reply tutorial/opt/aero/plot_optHist.py:18]: Agree with this

File: tutorial/opt/aero/plot_optHist.py
Comment (marcomangano) [reply tutorial/opt/aero/plot_optHist.py:29]: This is also a good idea

File: tutorial/opt/aero/plot_optHist.py
Comment (marcomangano) [reply tutorial/opt/aero/plot_optHist.py:22]: See my comment above, we could check and break out of main() or just return an error, but I honestly don't have strong opinions.
I am more concerned with the failing tests now but this needs a quick cleanup before approval

Comment (ewu63) [reply tutorial/opt/aero/plot_optHist.py:22]: Try/excepts are pretty non-pythonic and we should avoid them if possible. It should just be a simple check of if the file exists, and error out if not.

**Comment by umarsajjad1992** [comment]
I was thinking the same thing. Will implement a check and break out of main loop. Umar SajjadOn Jun 7, 2024, at 9:40 PM, Marco Mangano ***@***.***> wrote:﻿
@marcomangano commented on this pull request.



In tutorial/opt/aero/plot_optHist.py:
>  
-axes[0].plot("nMajor", "optimality", data=histValues, label="Optimality")
-axes[0].plot("nMajor", "feasibility", data=histValues, label="Feasibility")
-axes[0].set_yscale("log")
-axes[0].axhline(1e-4, linestyle="--", color="gray")
-axes[0].annotate("Convergence Criteria", xy=(3, 9e-5), ha="left", va="top", fontsize=24, color="gray")
-axes[0].legend(fontsize=20, labelcolor="linecolor", loc="upper right", frameon=False)
-axes[0].autoscale(enable=True, tight=True)
+    try:

See my comment above, we could check and break out of main() or just return an error, but I honestly don't have strong opinions.
I am more concerned with the failing tests now but this needs a quick cleanup before approval

File: tutorial/opt/aero/plot_optHist.py
Comment (umarsajjad1992) [reply tutorial/opt/aero/plot_optHist.py]: I have removed the try/excepts block for a simple os.path.isfile check

File: tutorial/opt/aero/plot_optHist.py
Comment (umarsajjad1992) [reply tutorial/opt/aero/plot_optHist.py]: I have removed the plt.rcParams and parser lines of code outside the main function

File: tutorial/opt/aero/plot_optHist.py
Comment (umarsajjad1992) [reply tutorial/opt/aero/plot_optHist.py]: I have removed the args.opt from args parser and instead incorporated hist.getMetadata()['optimizer'] to check for the optimizer

File: tutorial/opt/aero/plot_optHist.py
Comment (umarsajjad1992) [reply tutorial/opt/aero/plot_optHist.py]: I have removed code duplication by:
    if SNOPT:
        # make 5 subplots
        # plot optimality/feasibility on axes[4]
    else:
        # make 4 subplots
    # plot the rest on axes[-4:]

**Review by ewu63 — APPROVED** [review]

**Comment by ewu63** [comment]
I'm pretty sure the failure is just because we're running SNOPT tests and this is a PR submitted by an external contributor without access to that optimizer (hence this PR...ironically). May want to add skiptest or something similar to what we do with pyOptSparse tests.

**Comment by marcomangano** [comment]
> I'm pretty sure the failure is just because we're running SNOPT tests and this is a PR submitted by an external contributor without access to that optimizer (hence this PR...ironically). May want to add skiptest or something similar to what we do with pyOptSparse tests.

You were right, the ESP test was missing the `skiptest`. Merging the PR once I verify that the tests pass.

**Review by marcomangano — APPROVED** [review]

File: tutorial/opt/aero/plot_optHist.py
Comment (ewu63) [inline tutorial/opt/aero/plot_optHist.py:58]: I generally prefer having all `plt` commands before any function definition, at the module level. Not blocking this PR though, since this is probably personal preference.

**Review by ewu63 — APPROVED** [review]

**Pull request merged**
